### PR TITLE
Make tree internal to variant and seek to site

### PR DESCRIPTION
**Pull request by benjeffery**
Part of #605

### CODE REVIEW

**Comment by benjeffery** [comment]
There's an issue here with the lifetime of the `site` object as the `variant` stores a reference to it. This wasn't a problem before as the `site` was owned by the `tree`. 
I'm going to fix this by having the `site` owned by `variant` and changing the argument from a `site` object to being a site id.

**Comment by jeromekelleher** [comment]
I think we should make the argument to ``tsk_variant_decode`` a site ID, if only because it makes the client code simpler:
```c
tsk_variant_t var;
tsk_id_t site_id;

for (site_id = 0; site_id < (tsk_id_t) tsk_treeseq_get_num_sites(ts); site_id++) {
     ret = tsk_variant_decode(&var, site_id, 0);
     if (ret != 0) {
          goto out;
     }
} 
```
vs
```c
tsk_variant_t var;
tsk_site_t site;
tsk_id_t site_id;

for (site_id = 0; site_id < (tsk_id_t) tsk_treeseq_get_num_sites(ts); site_id++) {
     ret = tsk_treeseq_get_site(ts, site_id, &site);
     if (ret != 0) {
          goto out;
     }
     ret = tsk_variant_decode(&var, &site, 0);
     if (ret != 0) {
          goto out;
     }
} 
```

It's not like we save anything by having to call ``tsk_treeseq_get_site`` externally, and we'd actually have to work harder to validate the ``tsk_site_t`` object passed in as well (what tree sequence is it for? Is it some kind of random garbage the user has passed in?). 

The first way is much neater I think.

**Comment by jeromekelleher** [comment]
Why not keep storing the site on the variant object @benjeffery? We're keeping a reference to the underlying tree sequence, so there's basically no overhead in storing a pointer to a site, is there?

**Comment by benjeffery** [comment]
> Why not keep storing the site on the variant object @benjeffery? We're keeping a reference to the underlying tree sequence, so there's basically no overhead in storing a pointer to a site, is there?

I started the work to make variant lookup and store the site when I realised we didn't actually need to store it. I'm in two minds about it though as the site lookup is cheaper than I realised, basically copying a few pointers. Thought I would upload this version for feedback. If you feel strongly to store the site then we can switch to that, there was more breakage here than I expected so probably that is the right thing to do.

**Comment by jeromekelleher** [comment]
Since it's cheap I don't think it's worth the breakage. From a client perspective, it's nice to be able to do things like:
```c
tsk_variant_t var;
tsk_id_t site_id;

for (site_id = 0; site_id < (tsk_id_t) tsk_treeseq_get_num_sites(ts); site_id++) {
     ret = tsk_variant_decode(&var, site_id, 0);
     if (ret != 0) {
          goto out;
     }
     printf("Variant at site %d (pos=%f)\n", var.site.id, var.site.position);
} 
```
Otherwise, you have to do the site lookup again in the client code.

File: c/tskit/genotypes.c [hunk 1]
```diff
@@ -157,6 +141,17 @@ tsk_variant_init(tsk_variant_t *self, const tsk_treeseq_t *tree_sequence,
     tsk_size_t num_samples_alloc;
 
     tsk_memset(self, 0, sizeof(tsk_variant_t));
+    self->site = malloc(sizeof(*self->site));
```
Comment (jeromekelleher) [inline c/tskit/genotypes.c:144]: I think we could probably avoid this malloc, but let's not change it yet to keep the diff size managable. I've made  #2161 to track.

**Review by jeromekelleher — APPROVED** [review]
LGTM!

File: c/tskit/genotypes.c [hunk 2]
```diff
@@ -485,12 +470,19 @@ tsk_variant_update_site(tsk_variant_t *self, tsk_tree_t *tree, const tsk_site_t
...
+
+    ret = tsk_tree_seek(&self->tree, self->site->position, 0);
+    if (ret != 0) {
+        goto out;
+    }
 
     /* When we have a no specified samples we need sample lists to be active
      * on the tree, as indicated by the presence of left_sample */
-    if (!by_traversal && tree->left_sample == NULL) {
+    if (!by_traversal && self->tree.left_sample == NULL) {
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply c/tskit/genotypes.c:485]: Yes, will remove.

**Pull request merged**